File: backend/GeoTrack.Api/Models/Asset.cs
using NetTopologySuite.Geometries;

namespace GeoTrack.Api.Models;

public class Asset
{
    public Guid Id { get; set; }

    public required string Name { get; set; }

    public required string Description { get; set; }

    public required string AssetType { get; set; }

    /// <summary>
    /// Geographic location stored as PostGIS Point (SRID 4326 - WGS84)
    /// </summary>
    public required Point Location { get; set; }

    public AssetStatus Status { get; set; } = AssetStatus.Active;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Optional metadata stored as JSON
    /// </summary>
    public string? Metadata { get; set; }
}

public enum AssetStatus
{
    Active,
    Inactive,
    Maintenance,
    Retired
}

File: frontend/src/types/Asset.ts
export interface Asset {
  id: string;
  name: string;
  description: string;
  assetType: string;
  latitude: number;
  longitude: number;
  status: AssetStatus;
  createdAt: string;
  updatedAt: string;
  metadata?: string;
}

export enum AssetStatus {
  Active = 'ACTIVE',
  Inactive = 'INACTIVE',
  Maintenance = 'MAINTENANCE',
  Retired = 'RETIRED'
}

export interface GeoFence {
  id: string;
  name: string;
  description?: string;
  boundary: GeoJsonPolygon;
  type: GeoFenceType;
  isActive: boolean;
  createdAt: string;
}

export enum GeoFenceType {
  Inclusion = 'INCLUSION',
  Exclusion = 'EXCLUSION'
}

export interface GeoJsonPolygon {
  type: 'Polygon';
  coordinates: number[][][];
}

export interface Coordinate {
  latitude: number;
  longitude: number;
}

export interface CreateAssetInput {
  name: string;
  description: string;
  assetType: string;
  latitude: number;
  longitude: number;
  status?: AssetStatus;
  metadata?: string;
}

export interface UpdateAssetInput {
  id: string;
  name?: string;
  description?: string;
  assetType?: string;
  latitude?: number;
  longitude?: number;
  status?: AssetStatus;
  metadata?: string;
}

export interface UpdateAssetLocationInput {
  id: string;
  latitude: number;
  longitude: number;
}

export interface NearbySearchInput {
  latitude: number;
  longitude: number;
  distanceMeters: number;
}

export interface AssetMarkerProps {
  asset: Asset;
  onClick?: (asset: Asset) => void;
  isSelected?: boolean;
}

export interface MapViewProps {
  assets: Asset[];
  geoFences?: GeoFence[];
  center?: [number, number];
  zoom?: number;
  onAssetClick?: (asset: Asset) => void;
  selectedAssetId?: string;
}
